privacy/ccpa/policy.go

```diff
@@ -25,174 +22,103 @@ func ReadFromRequest(req *openrtb2.BidRequest) (Policy, error) {
 	}
 
 	// Read consent from request.regs.ext
-	if req.Regs != nil && len(req.Regs.Ext) > 0 {
-		var ext openrtb_ext.ExtRegs
-		if err := json.Unmarshal(req.Regs.Ext, &ext); err != nil {
-			return Policy{}, fmt.Errorf("error reading request.regs.ext: %s", err)
-		}
-		consent = ext.USPrivacy
+	err := req.ExtractRegExt()
+	if err != nil {
+		return Policy{}, fmt.Errorf("error reading request.regs.ext: %s", err)
+	}
+	if req.RegExt != nil {
+		consent = req.RegExt.GetUSPrivacy()
 	}
-
 	// Read no sale bidders from request.ext.prebid
-	if len(req.Ext) > 0 {
-		var ext openrtb_ext.ExtRequest
-		if err := json.Unmarshal(req.Ext, &ext); err != nil {
-			return Policy{}, fmt.Errorf("error reading request.ext.prebid: %s", err)
-		}
-		noSaleBidders = ext.Prebid.NoSale
+	err = req.ExtractRequestExt()
+	if err != nil {
+		return Policy{}, fmt.Errorf("error reading request.ext: %s", err)
+	}
+	reqPrebid := req.RequestExt.GetPrebid()
+	if reqPrebid != nil {
+		noSaleBidders = reqPrebid.NoSale
 	}
 
 	return Policy{consent, noSaleBidders}, nil
 }
 
 // Write mutates an OpenRTB bid request with the CCPA regulatory information.
-func (p Policy) Write(req *openrtb2.BidRequest) error {
+func (p Policy) Write(req *openrtb_ext.RequestWrapper) error {
```

Comment:
The intent of this method is to perform no changes to the request if there is an error. Now that the only errors come from the Extract calls, I recommend moving those to the top in a group:

```
if err := req.ExtractRegExt(); err != nil {
	return err
}
if err := req.ExtractRequestExt(); err != nil {
	return err
}
```

Reply:
The current setup follows the old flow, where if the Regs extracted correctly, but the request.Ext failed, then the buildRegs() function will have been called before erroring out. Your proposed change will change the behavior on certain errors from the behavior before this PR. We can do that if you believe that is the correct behavior, but I did not consider changing this behavior when writing this code.

Reply:
Done